sympy/polys/matrices/sdm.py

```diff
@@ -31,6 +31,32 @@ def __init__(self, elemsdict, shape, domain):
         if not all(0 <= c < n for row in self.values() for c in row):
             raise DDMBadInputError("Column out of range")
 
+    def getitem(self, i, j):
+        try:
+            return self[i][j]
+        except KeyError:
+            m, n = self.shape
+            if -m <= i < m and -n <= j < n:
+                try:
+                    return self[i % m][j % n]
+                except KeyError:
+                    return self.domain.zero
+            else:
+                raise IndexError("index out of range")
+
+    def extract_slice(self, slice1, slice2):
+        m, n = self.shape
+        ri = range(m)[slice1]
+        ci = range(n)[slice2]
+
+        sdm = {}
+        for i, row in self.items():
+            if row:
+                if i in ri:
+                    sdm[ri.index(i)] = {ci.index(j): e for j, e in row.items() if j in ci}
```

Comment:
This should check for an empty row.

Reply:
I am extremely sorry, I thought I already fixed it.
Thanks for pointing it out.

Reply:
@oscarbenjamin, this has been already fixed:
```py
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix

In [2]: dM = DomainMatrix({1: {0: 1}}, (2,1), ZZ)

In [3]: dM[0:, 0]
Out[3]: DomainMatrix({1: {0: 1}}, (2, 1), ZZ)

In [4]: dM.to_Matrix()[0:, 0]
Out[4]:
⎡0⎤
⎢ ⎥
⎣1⎦

In [5]: dM = DomainMatrix({2: {2: ZZ(1)}, 4: {4: ZZ(1)}}, (5, 5), ZZ)

In [6]: dM[:2,:2]
Out[6]: DomainMatrix({}, (2, 2), ZZ)

In [7]: dM.to_Matrix()[:2,:2]
Out[7]:
⎡0  0⎤
⎢    ⎥
⎣0  0⎦

In [8]: dM[2:,2:]
Out[8]: DomainMatrix({0: {0: 1}, 2: {2: 1}}, (3, 3), ZZ)

In [9]: dM.to_Matrix()[2:,2:]
Out[9]:
⎡1  0  0⎤
⎢       ⎥
⎢0  0  0⎥
⎢       ⎥
⎣0  0  1⎦

In [10]: dM[2:,2:].to_Matrix()
Out[10]:
⎡1  0  0⎤
⎢       ⎥
⎢0  0  0⎥
⎢       ⎥
⎣0  0  1⎦

```
You can also refer to the [tests](https://github.com/sympy/sympy/pull/21118/files#diff-91b314a98348bd4eb4e29009fffda338d594c2e4f9169cf276b198e7c9bd40a7R603-R610) which checks it.
Thanks.

Reply:
What we need to see is not how it looks when converted to a Matrix but whether the SDM has an empty dic in its internal representation i.e. `dict(dM.rep)`.

Reply:
@oscarbenjamin, Kindly review the latest commit.

After the latest commit:
```py
In [2]: dM = DomainMatrix({2: {2: ZZ(1)}, 4: {4: ZZ(1)}}, (5, 5), ZZ)

In [3]: dict(dM[2:, 6:].rep)
Out[3]: {}
```
Before the commit:
```py
In [2]: dM = DomainMatrix({2: {2: ZZ(1)}, 4: {4: ZZ(1)}}, (5, 5), ZZ)

In [3]: dict(dM[2:, 6:].rep)
Out[3]: {0: {}, 2: {}}
```
Thanks.